### PR TITLE
completed treraform file

### PR DESCRIPTION
**Pull request by talexandru1987**
Please make sure you are raising this PR against your own repository and not the original. If it says https://github.com/CorndelWithSoftwire in the search bar right now, you are in the wrong place!

Please also don't forget to submit the PR's url to Aptem afterwards (by clicking on the corresponding exercise component, pasting the url in the textbox that appears and then pressing the finish button).

### CODE REVIEW

File: .github/workflows/my-ci-pipeline.yml
Comment (JackMeadDev) [inline .github/workflows/my-ci-pipeline.yml]: Just an init? Are we intentionally not handling `terraform apply` here?

File: main.tf
Comment (JackMeadDev) [inline main.tf]: Probably not the best name for this container, but reasonable to re-use the one from this workshop I suppose!

File: main.tf
Comment (JackMeadDev) [inline main.tf]: This should actually match your app! And you may also see that this syntax is deprecated, so TF would prefer:
```terraform
    docker_image_name     = "appsvcsample/python-helloworld:latest"
    docker_registry_url   = "https://index.docker.io"
```
(But with your image name!)

File: main.tf
Comment (JackMeadDev) [inline main.tf]: If you want these extra outputs, I'd argue for keeping them in the `outputs.tf` file!

File: variables.tf
Comment (JackMeadDev) [inline variables.tf:5]: Nice addition 😄 One alternative I've seen (or maybe used, but I can't remember why!) is to explicitly restrict it to a preset list, using the `contains` rule; something like:
```
  validation {
    condition     = contains(["valid_prefix_1", "prefix_2", "prefix_3"], var.prefix)
    error_message = "Valid values for var: prefix are (...)."
  } 
```

File: variables.tf
Comment (JackMeadDev) [inline variables.tf]: This doesn't match your new app service name I think!

**Review by JackMeadDev — COMMENTED** [review]
This looks good in structure, but I can see a few things that will block it from actually working - not least that the deployed service currently just [has a default Python hello-world page!](https://cohort28-terraformtodo.azurewebsites.net/)

Hopefully it's mostly quick changes, but it would be good to see the app actually deployed to this new infra before we merge

File: main.tf
Comment (talexandru1987) [reply main.tf]: Done :)

File: main.tf
Comment (talexandru1987) [reply main.tf]: Done

File: main.tf
Comment (talexandru1987) [reply main.tf]: I have moved them to my outputs.tf file

File: variables.tf
Comment (talexandru1987) [reply variables.tf]: Done

File: .github/workflows/my-ci-pipeline.yml
Comment (JackMeadDev) [inline .github/workflows/my-ci-pipeline.yml]: This definitely won't be enough to work in isolation!
I'm expecting to see:
* Secrets loaded into Environment variables for `ARM_CLIENT_ID, ARM_CLIENT_SECRET, ARM_TENANT_ID, ARM_SUBSCRIPTION_ID`
* Secrets/variables for each of your non-default input variables
* `-auto-approve`, otherwise TF will just sit here and spin waiting for confirmation
* The `curl` command wants to execute against the terraform output so that we don't need to specify the Hook address directly

File: main.tf
Comment (JackMeadDev) [inline main.tf]: Your existing app service uses `talexandru87/todo-app:prod` - is this intentionally different?

**Review by JackMeadDev — APPROVED** [review]
Thanks for making the changes - this looks good although I can see the latest changes haven't actually been applied so we'll have to find out if they work, and I think the pipeline will still need a few tweaks to be passing! Happy for you to go ahead and merge when you're ready to experiment with the pipeline, or if it's not clear what I mean then reach out by email and we can set up a call to chat through this 🙂 

Either way, once you're happy that's working, then go ahead and delete the old non-IaC resources!

File: .github/workflows/my-ci-pipeline.yml
Comment (talexandru1987) [reply .github/workflows/my-ci-pipeline.yml]: are you free for a call this week, I have a few questions about this?

File: main.tf
Comment (JackMeadDev) [inline main.tf:54]: I know we saw issues trying to use the generated Mongo connection string so happy that you stuck with the variable here, but I managed to track down that resolving the deprecation warning seemed to resolve the permissions issue too for someone else, so if you're interested we could try switching out the `var.mongodb_connection_string` for `azurerm_cosmosdb_account.terraCosmos.primary_mongodb_connection_string`

File: todo_app/tests/test_view_model.py
Comment (JackMeadDev) [inline todo_app/tests/test_view_model.py:76]: Not particularly relevant here but all this commented out code is very ugly - worth noting that we can tag them with `@pytest.mark.skip()` to temporarily disable them in a slightly more visible way (each run will note that it was skipped, which is more likely to prompt someone to update them than simply commenting them out!)

File: .github/workflows/my-ci-pipeline.yml
Comment (JackMeadDev) [inline .github/workflows/my-ci-pipeline.yml:83]: You'll need to do this stage _before_ the TF init, since that also relies on having permission to speak to Azure.

That said, personally I prefer just specifying them in the env block, e.g. with:
```
    env:
      ARM_CLIENT_ID: ${{ secrets.ARM_CLIENT_ID }}
      ARM_CLIENT_SECRET=${{ secrets.ARM_CLIENT_SECRET }}
      ARM_TENANT_ID=${{ secrets.ARM_TENANT_ID }}
      ARM_SUBSCRIPTION_ID=${{ secrets.ARM_SUBSCRIPTION_ID }}
      TF_VAR_prefix: "cohort28"
```

(That said, are there not other variables we need to specify here? E.g. oauth IDs and secrets?)

**Review by JackMeadDev — COMMENTED** [review]
Happy this is merged, but note that it's still not passing in your pipeline! I've left a comment on the key issue, the environment variables, and where I think you might have missed later things, but this looks really close, so will sign off and we can tweak up the final bits in the future exercises 😄

**Pull request merged**
